File: ts/src/path.ts
/* Copyright (c) 2022-2026 Richard Rodger, MIT License */

import { Tabnas, Plugin, Rule, Context } from '@tabnas/parser'

type PathOptions = {}

// The host-grammar rules the plugin hooks. The Tabnas engine ships no
// grammar of its own, so these names must match the rules supplied by
// whatever grammar plugin the consumer installs. The standard
// value/map/pair/list/elem rule set uses these names. Each empty rule
// entry causes tabnas.grammar() to auto-wire any matching
// @<rulename>-<phase> function refs as state actions.
const HOOKED_RULES = ['val', 'map', 'pair', 'list', 'elem']

// Preallocated array pool for path tracking. Each depth level gets
// a reusable array of that length. The arrays are mutated in place
// as the parser traverses the tree — no allocation per pair/elem.
//
// IMPORTANT: r.k.path is a mutable, shared array. Client code that
// needs to retain the path beyond the current parse callback MUST
// copy it (e.g. r.k.path.slice() or [...r.k.path]).
const MAX_PATH_DEPTH = 64
const pathPool: any[][] = []
for (let i = 0; i <= MAX_PATH_DEPTH; i++) {
  pathPool[i] = new Array(i)
}


/* Keeps track of the property path to the current location.
 * Example: {a:{b:1 ## path=["a","b"]
 * Use the Rule.k key-value store so that the path is propagated to children and followers.
 * Depth must be greater than 0 - ensures path only starts once top level implicit is set up.
 */
const Path: Plugin = (tabnas: Tabnas, _options: PathOptions) => {
  const refs: Record<string, Function> = {
    '@val-bo': (r: Rule, ctx: Context) => {
      // At top level, create path array, or inherit from meta context.
      if (0 === r.d) {
        const base = ctx.meta.path?.base
        if (base && base.length > 0) {
          const arr = pathPool[base.length] || new Array(base.length)
          for (let i = 0; i < base.length; i++) arr[i] = base[i]
          r.k.path = arr
          r.k.pathDepth = base.length
        }
        else {
          r.k.path = pathPool[0]
          r.k.pathDepth = 0
        }
      }
    },

    '@map-bo': (r: Rule) => {
      // Not in an array, so no need to track element index.
      r.k.index = undefined
    },

    '@pair-ao': (r: Rule) => {
      if (0 < r.d && r.u.pair) {
        const depth = (r.k.pathDepth || 0) + 1
        const arr = pathPool[depth] || (pathPool[depth] = new Array(depth))
        const parent = r.k.path
        const parentLen = depth - 1
        for (let i = 0; i < parentLen; i++) arr[i] = parent[i]
        arr[parentLen] = r.u.key
        r.child.k.path = arr
        r.child.k.pathDepth = depth
        r.child.k.key = r.u.key
      }
    },

    '@list-bo': (r: Rule) => {
      // In array, the path property is the element index.
      r.k.index = -1
    },

    '@elem-ao': (r: Rule) => {
      if (0 < r.d) {
        r.k.index = 1 + r.k.index
        const depth = (r.k.pathDepth || 0) + 1
        const arr = pathPool[depth] || (pathPool[depth] = new Array(depth))
        const parent = r.k.path
        const parentLen = depth - 1
        for (let i = 0; i < parentLen; i++) arr[i] = parent[i]
        arr[parentLen] = r.k.index
        r.child.k.path = arr
        r.child.k.pathDepth = depth
        r.child.k.key = r.k.index
        r.child.k.index = r.k.index
      }
    },
  }

  // Declare the rules to hook and attach the refs. The empty rule specs
  // leave the host grammar's rules intact; only the @<rule>-<phase> refs
  // are wired in as state actions.
  const grammarDef: any = {
    rule: Object.fromEntries(HOOKED_RULES.map((name) => [name, {}])),
    ref: refs,
  }
  tabnas.grammar(grammarDef)
}

Path.defaults = {} as PathOptions

export { Path }

export type { PathOptions }

File: go/path.go
/* Copyright (c) 2022-2026 Richard Rodger and other contributors, MIT License */

package tabnaspath

import (
	tabnas "github.com/tabnas/parser/go"
)

const Version = "0.2.1"

type PathOptions struct{}

// hookedRules are the host-grammar rules the plugin attaches to. The
// Tabnas engine ships no grammar of its own, so these names must match
// the rules supplied by whatever grammar plugin the consumer installs.
// The standard value/map/pair/list/elem rule set uses these names.
var hookedRules = []string{"val", "map", "pair", "list", "elem"}

// Path is a Tabnas plugin that tracks the property path to the current
// location during parsing. The path is stored in Rule.K["path"] as a
// []any slice of string keys and int indices.
func Path(j *tabnas.Tabnas, opts map[string]any) error {
	refs := map[tabnas.FuncRef]any{
		"@val-bo": tabnas.StateAction(func(r *tabnas.Rule, ctx *tabnas.Context) {
			if r.D == 0 {
				var base []any
				if ctx.Meta != nil {
					if pm, ok := ctx.Meta["path"].(map[string]any); ok {
						if b, ok := pm["base"].([]any); ok {
							base = make([]any, len(b))
							copy(base, b)
						}
					}
				}
				if base == nil {
					base = []any{}
				}
				r.EnsureK()["path"] = base
			}
		}),

		"@map-bo": tabnas.StateAction(func(r *tabnas.Rule, ctx *tabnas.Context) {
			delete(r.K, "index")
		}),

		"@pair-ao": tabnas.StateAction(func(r *tabnas.Rule, ctx *tabnas.Context) {
			if r.D > 0 && r.U["pair"] != nil {
				key := r.U["key"]
				parentPath := toPathSlice(r.K["path"])
				childPath := make([]any, len(parentPath)+1)
				copy(childPath, parentPath)
				childPath[len(parentPath)] = key
				ck := r.Child.EnsureK()
				ck["path"] = childPath
				ck["key"] = key
			}
		}),

		"@list-bo": tabnas.StateAction(func(r *tabnas.Rule, ctx *tabnas.Context) {
			r.EnsureK()["index"] = -1
		}),

		"@elem-ao": tabnas.StateAction(func(r *tabnas.Rule, ctx *tabnas.Context) {
			if r.D > 0 {
				idx := 0
				if v, ok := r.K["index"].(int); ok {
					idx = v + 1
				}
				r.EnsureK()["index"] = idx
				parentPath := toPathSlice(r.K["path"])
				childPath := make([]any, len(parentPath)+1)
				copy(childPath, parentPath)
				childPath[len(parentPath)] = idx
				ck := r.Child.EnsureK()
				ck["path"] = childPath
				ck["key"] = idx
				ck["index"] = idx
			}
		}),
	}

	// Declare the rules to hook. Each empty rule spec causes Grammar to
	// auto-wire the matching @<rulename>-<phase> refs above as state
	// actions, without otherwise altering the host grammar's rules.
	gsRule := make(map[string]*tabnas.GrammarRuleSpec, len(hookedRules))
	for _, name := range hookedRules {
		gsRule[name] = &tabnas.GrammarRuleSpec{}
	}

	return j.Grammar(&tabnas.GrammarSpec{Ref: refs, Rule: gsRule})
}

func toPathSlice(v any) []any {
	if p, ok := v.([]any); ok {
		return p
	}
	return []any{}
}
